src/box/txn_limbo.h

```diff
@@ -280,6 +280,16 @@ txn_limbo_is_replica_outdated(struct txn_limbo *limbo, uint32_t replica_id)
 	return res;
 }
 
+/** Return id of the node owning the synchro queue. */
+static inline uint32_t
+txn_limbo_owner_id(struct txn_limbo *limbo)
+{
+	latch_lock(&limbo->promote_latch);
+	uint32_t owner_id = limbo->owner_id;
+	latch_unlock(&limbo->promote_latch);
+	return owner_id;
+}
+
```

Comment:
> The completely correct fix would be to write the term the transaction
was written in with each transaction and replace
txn_limbo_is_replica_outdated with txn_limbo_is_request_outdated, so
that we decide whether to filter the request or not judging by the term
it was applied in, not by the term we seen in some past PROMOTE from the
node. This fix seems too costy though, given that we only miss one case
of split-brain at the moment when the user enables master-master
replication (by writing a DEMOTE). And in master-master there is no such
thing as a split-brain.

Not sure I understand the problem. Writing term with each txn is the same as writing only term change event. You just 'compact' all duplicated term values. When you receive data from a replica, you always know whether it did apply the latest demote or not. Why doesn't it allow to reject the old leader? You know which replica applied which term last. And you know the term of demote - if your limbo owner == 0, then the demote term is `promote_greatest_term`. Hence if you got data from an old leader, you will rightfully reject it by `promote_greatest_term`. We update `promote_greatest_term` even for demotes.

There certainly should be a simpler explanation and a solution to what is happening in the ticket. Or do I miss something obvious?

Reply:
> When you receive data from a replica, you always know whether it did apply the latest demote or not.

That's exactly what I thought, but if I'm not confused here, that's not always true.

You see, in `promote_term_map` we only track terms of PROMOTE/DEMOTE **authored** by a specific node.

So, when you have 3 nodes, PROMOTE node 2 in term 1, PROMOTE node 3 in term 2, PROMOTE node 1 in term 3, `promote_term_map` will look like: `id_1: 3, id_2: 1, id_3: 2`. Even when nodes 2 and 3 receive PROMOTE from node 1 in term 3.
This is okay for filtering after promotions. Because we allow everyone to send us data authored by node 1 in term 3. As long as term 3 is the biggest term.
But this is not fine for demotions. After a DEMOTE node 1 term 4, everyone should be writeable, but `promote_term_map` will look like `id_1: 4, id_2: 1, id_3: 2`. So any data coming not from node1 will be considered malicious.

After understanding this, I thought we might track the biggest term **received** by the node, and not **authored** by it.
But this won't work: imagine node 1 writes DEMOTE node 1 term 4. How is it supposed to know that nodes 2 and 3 received this request? It will never see this request again, because it has already applied it. It will be filtered out by vclock.

So having once applied a PROMOTE/DEMOTE request, we will skip it in replication stream from everyone else. And AFAICS it's impossible to know which node applied which PROMOTE/DEMOTE.

Now you might say we may track PROMOTE/DEMOTE requests even before filtering them by vclock. But this won't work either: if both node 1 and node 2 applied some old PROMOTE long time ago, this PROMOTE won't be present in xlogs of either of them. Imagine node 2 stays in the term of that old PROMOTE and writes a lot of data in the old term. Since the PROMOTE isn't in xlogs anymore (they are already deleted), node1 knows nothing of the term the data is written in.

This all was fine for PROMOTE, because we always know who has done the latest promote and only apply his data. But this is not enough for demote. Because we need to know about every server whether it has applied the demote or not.

I fear my explanations are often a mess. So please let me know if we should discuss this verbally.

Do you want me to add some kind of explanation like this to the commit message?

Reply:
Thanks for the explanation, I see the problem now indeed. In that case the solution in the PR looks fine so far.
As for the proper solution - sure, term per txn would be better. But not perfect still. Because some bad node could emit the txn in the demoted term but before actually receiving the `DEMOTE` itself. It is possible? There is a gap between term bump and `DEMOTE` with that term, right?

The truly proper solution would be to solve this issue: be able to tell for each node its latest applied request type (`PROMOTE` or `DEMOTE`) *and* its term.

For that we could introduce `PROMOTE_ACK`(the name is a bit unfortunate though). Node could write it into WAL under its own vclock in the same WAL write batch as `PROMOTE`/`DEMOTE` (or not same batch - just in the same limbo txn then). It would be replicated and the other nodes would be able to tell whether this node tried to emit any transactions after knowing about demote or not. It allows not to write term per txn and still be able to tell which transaction from a given node was emitted in which term and whether it was after or before `PROMOTE`/`DEMOTE`.

`PROMOTE_ACK` would contain acked request type (`PROMOTE`/`DEMOTE`) and term. I tried to figure out whether we can just rely on vclock, but it is more complicated. By another node's vclock we can't tell how much did it progress in our own vclock component. We would need matrix clock for that, which is a huge overkill.

I also thought about doing it over piggybacking more info in applier acks or just sending new messages or reusing raft messages, but it wouldn't work for non-fullmesh (which should work fine even in some synchronous replication cases and definitely in all non-synchronous replication cases).

Wdyt about this? Might be complicated to do, so I am fine with the current patch as well.

Reply:
> Because some bad node could emit the txn in the demoted term but before actually receiving the DEMOTE itself. It is possible? There is a gap between term bump and DEMOTE with that term, right?

Yes. We could sign requests with *limbo term* instead of the raft one, I guess.
This would cause problems too, I think. While a new PROMOTE is written to WAL, we should somehow set its term to all the new requests, shouldn't we?

Looks like `PROMOTE_ACK` would fix the issue, indeed. And we would always know which part of a node's journal belongs to which term. And I think it would be best to write PROMOTE_ACK in the same wal batch, as the PROMOTE/DEMOTE itself. Then we close the gap between PROMOTE and PROMOTE_ACK write. And if PROMOTE write fails it will automatically abort the PROMOTE_ACK write.

This "knowing the exact term of each request" seems important, and we might need this in the future if some not yet discovered bug arises. But TBH I don't want to do this know.

> I also thought about doing it over piggybacking more info in applier acks or just sending new messages or reusing raft messages

AFAICS you would only be able to see if the node has applied the latest PROMOTE/DEMOTE or not.
But you still won't know the term of the currently replicated request. It might come from some old term.

Reply:
> We could sign requests with limbo term instead of the raft one, I guess.

Would this work?

Reply:
Most likely it would. But I hope this is theoretical - adding a new field to each txn would be too expensive IMO.